File: 123-best-time-to-buy-and-sell-stock-iii/123-best-time-to-buy-and-sell-stock-iii.cpp
class Solution {
public:
    int maxProfit(vector<int>& prices) {
        int n = prices.size(), mini = INT_MAX, maxi= INT_MIN;
        int dp1[n],dp2[n];
        memset(dp1,0,sizeof(dp1));
        memset(dp2,0,sizeof(dp2));
        for(int i = 0; i < n ;i++){
            mini = min(mini, prices[i]);
            if(i>0)
                dp1[i]= max(dp1[i-1],prices[i]-mini);
        }
        for(int i = n-1; i>=0; i--){
            maxi = max(maxi,prices[i]);
            if(i<n-1)
                dp2[i]= max(dp2[i],maxi- prices[i]);
        }
        int m =0;
        for(int i =0; i < n;i++)
            m = max(m, dp1[i]+dp2[i]);
        return m;
    }
};

File: 123-best-time-to-buy-and-sell-stock-iii/123-best-time-to-buy-and-sell-stock-iii.py
class Solution:
    def maxProfit(self, prices: List[int]) -> int:
        bs = [ 0 for _ in range(len(prices))]
        sb = [ 0 for _ in range(len(prices))]
        mini = prices[0]
        for idx, price in enumerate(prices):
            bs[idx] = max(price - mini, bs[idx-1] if idx > 0 else bs[idx])
            mini = min(mini, price)
        maxi = prices[-1]
        for idx in range(len(prices)-1, -1, -1):
            sb[idx] = max(maxi - prices[idx], sb[idx+1] if idx<len(prices)-1 else sb[idx])
            maxi = max(maxi, prices[idx])
        maxi = 0
        for idx in range(len(bs)):
            maxi = max(maxi, bs[idx]+sb[idx])
        return maxi
        
        
        
        
#BS BS
#SBS
#bbss
#sbsb


# [3,3,5,0,0,3,1,4]

# [0, 0, 2, 2, 2,2,2,2]
# [ 4  4  4  4  3  3 0]
# 0+4 =4
# 2+4=6